test/cpp/test_xla_sharding.cpp

```diff
@@ -239,5 +241,72 @@ TEST_F(XLAShardingTest, InputHandler) {
 }
 
+using TensorShardingStore = ShardingUtil::TensorShardingStore;
+
+TEST_F(XLAShardingTest, TensorShardingStore) {
+  TensorShardingStore::ShardingInfo info{
+      std::make_shared<XLATensor::ShardingSpec>(
+          xla::HloSharding::Replicate().ToProto()),
+      torch::lazy::Value()};
+
+  int handle = 1;
+  TensorShardingStore::RegisterShardingInfo(handle, info);
+  EXPECT_TRUE(TensorShardingStore::GetShardingInfo(0) == nullptr);
+  EXPECT_TRUE(TensorShardingStore::GetShardingInfo(handle) != nullptr);
+
+  TensorShardingStore::UnregisterShardingInfo(handle);
+  EXPECT_TRUE(TensorShardingStore::GetShardingInfo(handle) == nullptr);
+
+  TensorShardingStore::RegisterShardingInfo(handle, info);
+  TensorShardingStore::Reset();
+  EXPECT_TRUE(TensorShardingStore::GetShardingInfo(handle) == nullptr);
+}
+
+TEST_F(XLAShardingTest, ShardInputDataNodes) {
+  if ((xla::sys_util::GetEnvString(xla::env::kEnvPjRtDevice, "") == "") ||
+      (xla::ComputationClient::Get()->GetLocalDevices().size() < 2)) {
+    GTEST_SKIP()
+        << "`PJRT_DEVICE` is not set, with more than 2 local devices, ("
+        << xla::ComputationClient::Get()->GetLocalDevices().size()
+        << " local devices detected).";
+  }
+
+  at::Tensor tensor = at::ones({8}, at::TensorOptions(at::kFloat));
+  const torch::lazy::BackendDevice* device = GetDefaultDevice();
+  torch::lazy::Value data_node = torch::lazy::MakeNode<DeviceData>(
+      std::move(TensorToXlaData(tensor, *device)));
+  EXPECT_TRUE(xla::ComputationClient::Get()
+                  ->GetDataShards(UnwrapXlaData(
+                      torch::lazy::getBackend()->GetComputationDataFromNode(
+                          data_node.node.get())))
+                  .size() == 1);
+
+  // Takes a data node and returns a new node with sharded data handle.
+  XLATensor::ShardingSpecPtr sharding =
+      std::make_shared<XLATensor::ShardingSpec>(
+          xla::HloSharding::Replicate().ToProto());
+  torch::lazy::Value new_data_node =
+      ShardingUtil::ShardInputDataNodes(data_node, sharding);
+  EXPECT_TRUE(xla::ComputationClient::Get()
+                  ->GetDataShards(UnwrapXlaData(
+                      torch::lazy::getBackend()->GetComputationDataFromNode(
+                          new_data_node.node.get())))
+                  .size() > 1);
+
+  // Takes an op and propate sharding to its data node oeprands.
+  torch::lazy::NodePtr matmul = MatMul(data_node, data_node);
+  torch::lazy::Value new_matmul = ShardingUtil::ShardInputDataNodes(
+      torch::lazy::Value(matmul, 0), sharding);
+  XlaNode* xla_node = dynamic_cast<XlaNode*>(new_matmul.node.get());
+  for (int i = 0; i < xla_node->operands_as_nodes()->size(); ++i) {
+    torch::lazy::Node* node = xla_node->get_operand(i);
+    EXPECT_TRUE(
+        xla::ComputationClient::Get()
+            ->GetDataShards(UnwrapXlaData(
+                torch::lazy::getBackend()->GetComputationDataFromNode(node)))
+            .size() > 1);
```

Comment:
Hmm, I am still not entirely sure why do we need to replace the operand with sharded version if we shard the result node. I thought compile will take care of that?

Reply:
note that you might have
```
t1 = tensor(1)
t2 = t1 * 3
t3 = t1 * 2
t3 = shard(t3)
```
in this case I guess t2's operand is still the unsharded version?